src/ast/analysis/Type.cpp

```diff
@@ -823,23 +840,282 @@ void TypeAnalysis::print(std::ostream& os) const {
+            changed = true;
+        });
+
+        // Deduce numeric-constant polymorphism
+        auto setNumericConstantType = [&](const NumericConstant& nc, NumericConstant::Type ncType) {
+            if (contains(numericConstantType, &nc) && numericConstantType.at(&nc) == ncType) return;
+            changed = true;
+            numericConstantType[&nc] = ncType;
+        };
+
+        visitDepthFirst(program, [&](const NumericConstant& numericConstant) {
+            // Constant has a fixed type
+            if (numericConstant.getFixedType().has_value()) {
+                setNumericConstantType(numericConstant, numericConstant.getFixedType().value());
+                return;
+            }
+
+            // Otherwise, type should be inferred
+            TypeSet types = getTypes(&numericConstant);
+            auto hasOfKind = [&](TypeAttribute kind) -> bool {
+                return any_of(types, [&](const analysis::Type& type) { return isOfKind(type, kind); });
+            };
+            if (hasOfKind(TypeAttribute::Signed)) {
+                setNumericConstantType(numericConstant, NumericConstant::Type::Int);
+            } else if (hasOfKind(TypeAttribute::Unsigned)) {
+                setNumericConstantType(numericConstant, NumericConstant::Type::Uint);
+            } else if (hasOfKind(TypeAttribute::Float)) {
+                setNumericConstantType(numericConstant, NumericConstant::Type::Float);
+            } else {
+                // Type information no longer valid
+                if (contains(numericConstantType, &numericConstant)) {
+                    numericConstantType.erase(&numericConstant);
+                    changed = true;
+                }
+            }
+        });
+
+        // Deduce aggregator polymorphism
+        auto setAggregatorType = [&](const Aggregator& aggr, TypeAttribute attr) {
+            auto overloadedType = convertOverloadedAggregator(aggr.getBaseOperator(), attr);
+            if (contains(aggregatorType, &aggr) && aggregatorType.at(&aggr) == overloadedType) return;
+            changed = true;
+            aggregatorType[&aggr] = overloadedType;
+        };
+        visitDepthFirst(program, [&](const Aggregator& aggregator) {
+            if (isOverloadedAggregator(aggregator.getBaseOperator())) {
+                auto* targetExpression = aggregator.getTargetExpression();
+                if (isFloat(targetExpression)) {
+                    setAggregatorType(aggregator, TypeAttribute::Float);
+                } else if (isUnsigned(targetExpression)) {
+                    setAggregatorType(aggregator, TypeAttribute::Unsigned);
+                } else {
+                    setAggregatorType(aggregator, TypeAttribute::Signed);
+                }
+            } else {
+                if (contains(aggregatorType, &aggregator)) {
+                    assert(aggregatorType.at(&aggregator) == aggregator.getBaseOperator() &&
+                            "unexpected aggr type");
```

Comment:
The error message could be expanded a bit to be more meaningful - remember that it may be seen without context.